tests/s3_mock_server_tests.c

```diff
@@ -167,6 +167,206 @@ TEST_CASE(multipart_upload_mock_server) {
     return AWS_OP_SUCCESS;
 }
 
+/* Singleton used by tests in this file */
+static struct get_requests_header_tester {
+    struct aws_allocator *alloc;
+
+    /* Store the requests headers in the array. Array of struct aws_http_headers * */
+    struct aws_array_list headers_array;
+    struct aws_mutex lock;
+} s_get_requests_header_tester;
+
+static int s_get_requests_header_tester_init(struct aws_allocator *alloc) {
+    ASSERT_SUCCESS(aws_array_list_init_dynamic(
+        &s_get_requests_header_tester.headers_array, alloc, 1, sizeof(struct aws_http_headers *)));
+    ASSERT_SUCCESS(aws_mutex_init(&s_get_requests_header_tester.lock));
+    return AWS_OP_SUCCESS;
+}
+
+static void s_get_requests_header_tester_clean_up(void) {
+    /* iterate thought the headers array to clean up the headers */
+    for (size_t i = 0; i < aws_array_list_length(&s_get_requests_header_tester.headers_array); ++i) {
+        struct aws_http_headers *headers = NULL;
+        aws_array_list_get_at(&s_get_requests_header_tester.headers_array, &headers, i);
+        aws_http_headers_release(headers);
+    }
+    aws_mutex_clean_up(&s_get_requests_header_tester.lock);
+    aws_array_list_clean_up(&s_get_requests_header_tester.headers_array);
+}
+
+struct aws_http_stream *s_get_requests_header_make_request(
+    struct aws_http_connection *client_connection,
+    const struct aws_http_make_request_options *options) {
+    /**
+     * Record the headers in the array.
+     */
+    aws_mutex_lock(&s_get_requests_header_tester.lock);
+    struct aws_http_headers *headers = aws_http_message_get_headers(options->request);
+    /* Keep the headers alive until we clean up the tester. */
+    aws_http_headers_acquire(headers);
+    aws_array_list_push_back(&s_get_requests_header_tester.headers_array, &headers);
+    aws_mutex_unlock(&s_get_requests_header_tester.lock);
+
+    struct aws_http_stream *stream = aws_http_connection_make_request(client_connection, options);
+    return stream;
+}
+
+TEST_CASE(multipart_upload_unsigned_with_trailer_checksum_mock_server) {
+    (void)ctx;
+
+    struct aws_s3_tester tester;
+    ASSERT_SUCCESS(aws_s3_tester_init(allocator, &tester));
+    ASSERT_SUCCESS(s_get_requests_header_tester_init(allocator));
+
+    struct aws_s3_client_config client_config = {
+        .tls_mode = AWS_MR_TLS_DISABLED,
+    };
+
+    ASSERT_SUCCESS(aws_s3_tester_bind_client(&tester, &client_config, AWS_S3_TESTER_BIND_CLIENT_REGION));
+
+    struct aws_s3_client_vtable s3_client_get_requests_header_vtable = g_s3_client_default_vtable;
+    s3_client_get_requests_header_vtable.http_connection_make_request = s_get_requests_header_make_request;
+    struct aws_s3_client *client = aws_s3_client_new(allocator, &client_config);
+    ASSERT_NOT_NULL(client);
+    client->vtable = &s3_client_get_requests_header_vtable;
```

Comment:
An alternative to making the vtable global:

```suggestion
    struct aws_s3_client *client = aws_s3_client_new(allocator, &client_config);
    ASSERT_NOT_NULL(client);
    
    // Patch the vtable
    struct aws_s3_client_vtable s3_client_get_requests_header_vtable = *client->vtable;
    s3_client_get_requests_header_vtable.http_connection_make_request = s_get_requests_header_make_request;
    client->vtable = &s3_client_get_requests_header_vtable;
```